pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,350 @@
+	return &srv{
+		ctx:      ctx,
+		server:   server,
+		upgrader: &websocket.Upgrader{},
+	}
+}
+
+func (s *srv) RegisterRoutes(server *web.Server) {
+	group := server.Group(ttnpb.HTTPAPIPrefix + "/gs/io/basicstation")
+	group.GET("/discover", s.handleDiscover)
+	group.GET("/traffic/:id", s.handleTraffic)
+}
+
+func (s *srv) handleDiscover(c echo.Context) error {
+	ctx := c.Request().Context()
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "discover",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	_, data, err := ws.ReadMessage()
+	if err != nil {
+		logger.WithError(err).Debug("Failed to read message")
+		return err
+	}
+	var req messages.DiscoverQuery
+	if err := json.Unmarshal(data, &req); err != nil {
+		logger.WithError(err).Debug("Failed to parse discover query message")
+		return err
+	}
+
+	if req.EUI.IsZero() {
+		writeDiscoverError(s.ctx, ws, "Invalid request")
+		return errEmptyGatewayEUI
+	}
+
+	ids := ttnpb.GatewayIdentifiers{
+		EUI: &req.EUI.EUI64,
+	}
+	ctx, ids, err = s.server.FillGatewayContext(ctx, ids)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to fill gateway context")
+		writeDiscoverError(ctx, ws, "Router not provisioned")
+		return err
+	}
+
+	scheme := "ws"
+	if c.IsTLS() {
+		scheme = "wss"
+	}
+	res := messages.DiscoverResponse{
+		EUI: req.EUI,
+		Muxs: basicstation.EUI{
+			Prefix: "muxs",
+		},
+		URI: fmt.Sprintf("%s://%s%s", scheme, c.Request().Host, c.Echo().URI(s.handleTraffic, ids.GatewayID)),
+	}
+	data, err = json.Marshal(res)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to marshal response message")
+		writeDiscoverError(ctx, ws, "Router not provisioned")
+		return err
+	}
+	if err := ws.WriteMessage(websocket.TextMessage, data); err != nil {
+		logger.WithError(err).Warn("Failed to write discover response message")
+		return err
+	}
+	logger.Debug("Sent discover response message")
+	return nil
+}
+
+func (s *srv) handleTraffic(c echo.Context) error {
+	var latestUpstreamXTime int64
+	id := c.Param("id")
+	auth := c.Request().Header.Get(echo.HeaderAuthorization)
+	ctx := c.Request().Context()
+	if auth != "" {
+		md := metadata.New(map[string]string{
+			"id":            id,
+			"authorization": fmt.Sprintf("Bearer %s", auth),
+		})
+		if ctxMd, ok := metadata.FromIncomingContext(s.ctx); ok {
+			md = metadata.Join(ctxMd, md)
+		}
+		ctx = metadata.NewIncomingContext(s.ctx, md)
+	}
+
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "traffic",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+
+	ctx, ids, err := s.server.FillGatewayContext(ctx, ttnpb.GatewayIdentifiers{GatewayID: id})
+	if err != nil {
+		logger.WithError(err).Debug("Failed to fill gateway context")
+		return err
+	}
+
+	if auth == "" {
```

Comment:
I don't get this. If there is no auth, we just give the gateway link rights?
This would at least deserve a comment explaining why.

Reply:
Yes. Indeed. `FillGatewayContext` will assert the rights later down the road. For gateways without an auth but a valid EUI, it works the same as UDP (where we provide rights ourselves). Will add the comment.